### PR TITLE
Make session inaccessible files remembered (part 2/2)

### LINKED ISSUE #12079: Loading Current Session with missing files does not meet user expectation
### Description of the Issue

When you start Notepad++ and it loads the current session, if one or more of the files from the session are "missing" for any reason, those files aren't opened ... so when Notepad++ saves the new current session, those files are removed from the session.

If the files were truly permanently gone (deleted intentionally), that behavior makes sense.

But if the reason the files are gone is because of a temporary problem outside of their control, then the users get confused when those files don't come back into Notepad++ when the temporary problem is resolved.  Examples include:
1. the USB drive was bumped and Windows didn't see it was there, so Windows claims the files don't exist even though they are still on the USB stick
2. the network drive was temporarily unavailable or hadn't been fully connected when Notepad++ started looking for the files
3. the user ran in Admin mode temporarily (because run-as-admin doesn't see all mounted network drives)
4. wifi was temporarily unavailable so One Drive or DropBox or similar cloud folders weren't finished downloading files yet
5. any number of other temporary reasons that I couldn't think of off the top of my head

This isn't some hypothetical: Users have been complaining about this more, recently, in the Forum:
- https://community.notepad-plus-plus.org/topic/23380/files-from-sessions-are-lost-when-a-session-file-is-opened-and-not-connected-to-the-vpn-that-was-connected-but-dropped
- https://community.notepad-plus-plus.org/topic/23433/how-to-find-out-what-txt-files-were-open-in-the-last-session

### Steps to Reproduce the Issue

0. Start with new portable Notepad++
1. Open files `a.txt` and `b.txt`
2. Close Notepad++ so it saves the current session as session.xml
3. Move `b.txt` to `be.txt`
4. Open Notepad++: only `a.txt` is open, because `b.txt` doesn't exist
5. PROBLEM#1: `b.txt` is _not_ in **File > Recent Files...** list.
6. Close Notepad++ so it saves the current session as session.xml
7. Move `be.txt` back to `b.txt`
8. Open Notepad++: still only opens `a.txt`
9. PROBLEM#2: at this point, common users (as given by the two example topics above) expect that `b.txt` will also be open again; or, barring that, that there's at least a way to see what files used to be open (like the Recent Files History list).

### Desired Behavior

Based on my understanding, I think they would appreciate a two-part fix:
1. When session files are non-existent, they should be added to the "Recent Files History" so that Notepad++ has a record of files that used to be open but aren't any more.
2. If the **Preferences > Backup** settings could have an option called "Keep missing files in the saved session", which would default to ON.

The desired sequence of events:
1. Notepad++ launches
2. Parse the active session.xml
3. For each file, in session:
   - if file exists, open it
   - if file missing,
      1. add it to an internal list of "missing files"
      2. add it to the Recent Files History list
4. (normal editor usage)
5. User triggers closing Notepad++
6. Notepad++ checks "Keep missing files in the saved session" setting:
   - if ON, Notepad++ transfers the "missing files" list into the active session, along with the list of files that are really open
   - if OFF, Notepad++ only includes the files that still exist and are really open
7. Notepad++ closes
8. Next run
   - if option was ON at the last session-save, the files that were "missing" last time but exist again this time are re-opened, so the users files automatically come back once the network glitch is over
   - any files that were "missing" last time and are still missing this time are still listed in the Recent Files History list, so they have a record of what files still aren't reloaded.

Optional suggestion: Keep a missingCount attribute in the session.xml entry for each file.  If the file exists when the session is saved, that attribute is set to 0.  If the file was "missing" this run of Notepad++, add 1 for that file when the session.xml is saved.  When the session is loaded and missingCount is above some threshold (*), permanently delete it from the session file, rather than permanently keeping a deleted file in the session.  *: maybe hardcode the threshold at 5, or maybe make the threshold a setting in the preferences dialog, so that users could decide how many times to keep a missing file before finally giving up and removing it from the active session.

### Debug Information

```
Notepad++ v8.4.4   (64-bit)
Build time : Jul 15 2022 - 17:54:42
Path : C:\usr\local\apps\npp\npp.8.4.4.portable.x64\notepad++.exe
Command Line : a.txt b.txt
Admin mode : OFF
Local Conf mode : ON
Cloud Config : OFF
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 20H2
OS Build : 19042.1889
Current ANSI codepage : 1252
Plugins : 
    mimeTools (2.8)
    NppConverter (4.4)
    NppExport (0.4)
```

### LINKED ISSUE #12744: Notepad++ doesnt show an error message when attempting to load sessions with missing file paths
When loading a session with missing file paths, notepad++ provides no error message so the user has no idea why the missing files are not being loaded with the rest of the session. Would it be possible to add an enhancement so that notepad++ displays an error message with the list of files that it cannot load with the session?

### Debug Information
Notepad++ v8.4.8 (64-bit)
Build time : Dec 24 2022 - 19:40:27
Path : C:\Program Files\Notepad++\notepad++.exe
Command Line : "C:\Users\quest\Documents\test session"
Admin mode : ON
Local Conf mode : OFF
Cloud Config : OFF
OS Name : Windows 10 Home (64-bit)
OS Version : 22H2
OS Build : 19045.2364
Current ANSI codepage : 932
Plugins :
mimeTools (2.9)
NppConverter (4.5)
NppExport (0.4)
NPPJSONViewer (1.40)
NppToolBucket (1.10.6622.41516)

### LINKED ISSUE #13696: Restore opened file tabs for temporarly unavailable file paths
### Description of the Issue
When I forgot map network drive or mount crypted volume, or any else with logical hard drive letter is unavailable (E:\ F:\ to Z:\.) and run notepad++, then editor silently drops all opened files for that drive(s) without additional info, because paths are unavailable.

### Steps to Reproduce the Issue
1. in command line emulate network drive or crypted volume
1a. **run cmd.exe**
1b. **mkdir C:\W**  (subdir for emulation **W:\** drive, of course if **W:\** is free on your system)
1c. **subst W: C:\W**  (at now you have drive W  . please do not close cmd.exe, it will be helpful later
2. open notepad++ , create new tab for text file, and write any text inside this tab
3. write this file for example as **W:\testfile.txt**
4. close notepad++ and run it again. you will see opened file **testfile.txt** , this is OK
5. close notepad++
6. click again cmd.exe window, and delete emulated drive by run **subst W: /D**
7. open notepad++ again.
8. this is all. you will not be noticed about lost path to **W:\testfile.txt** , and tab will not be open again.

### Expected Behavior
I need some configuration for this one, by selecting between:
1. **silent** do not open unavailable paths
2. **warning window**, with info about unavailable paths, short info similar to 'please map network or crypto drives W:, X: etc.... and or select below...', with two buttons 'Do not open listed files' , and second 'Try again'.
3. continuation to ad.2 , if still any paths are unavailable, do corrected list again described at point 2

### Actual Behavior
currently default is silently abort opening tabs for unavailable paths

### Debug Information
There is no debug. This behavior exist on all available versions notepad++

### PR DESCRIPTION
**Pull request by donho**
Based from #14168
Continue with #14202

Fix #12079, fix #12744, fix #13696

### CODE REVIEW

**Comment by donho** [comment]
@Yaron10

Done, please check it.

**Comment by Yaron10** [comment]
@donho,

👍 
Thanks again for your work and patience.

I'll continue testing it tomorrow.
So far:

***

**STR:**
Open file `A`.
File -> Save Session: save it as SessionA.ses.
Close NPP.
Delete file `A`.
Open SessionA.ses from File-Explorer.
Click "Yes" in the two messages.

**Result:**
Two `A` tabs in the same view. 

***

**STR:**
Create two user-sessions.
Delete some file(s) from _one session file_.
Select the two user-sessions in File_explorer and open them in one go.

**Result:**
You can not know which session file has missing files.

How about replacing `bool loadSession(Session & session, bool isSnapshotMode = false, bool isUserCreatedSession = false);` with `bool loadSession(Session & session, bool isSnapshotMode = false, const TCHAR *sessionFileName = nullptr);`?

\- You could then display the user-session file-name in the message.

***

"Open Session in New Instance (and save on exit)" complicates the implementation/messages.

***

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/c8fafcd12b1f88c4817d7fee0a3f214d3772659c/PowerEditor/src/Notepad_plus.cpp#L6487

Is that still needed?

**Comment by donho** [comment]
@Yaron10 
> STR:
Open file A.
File -> Save Session: save it as SessionA.ses.
Close NPP.
Delete file A.
Open SessionA.ses from File-Explorer.
Click "Yes" in the two messages.

> Result:
Two A tabs in the same view.

Fixed.

> "Open Session in New Instance (and save on exit)" complicates the implementation/messages.

Could you elaborate it?

> Is that still needed?

Simplified.

Please check the latest commit.

File: PowerEditor/installer/nativeLang/french.xml [hunk 1]
```diff
@@ -1167,6 +1167,7 @@ Vous pouvez définir plusieurs marqueurs de colonne en utilisant un espace pour
 					<Item id="6821" name="secondes"/>
 					<Item id="6822" name="dans le dossier"/>
 					<Item id="6309" name="Mémoriser la session actuelle"/>
+					<Item id="6825" name="Mémoriser les fichiers inaccessibles de la session précédante"/>
```
Comment (Mateos81) [inline PowerEditor/installer/nativeLang/french.xml:1170]: Précédente*

File: PowerEditor/installer/nativeLang/french.xml [hunk 2]
```diff
@@ -1481,6 +1482,13 @@ Veuillez tester ces commandes et, si besoin, ré-éditer celles-ci.
 Autrement, vous pouvez revenir vers Notepad++ v8.5.2 et restaurer les données précédentes.
 Notepad++ va sauvegarder votre ancien &quot;shortcuts.xml&quot; et l'enregistrer sous &quot;shortcuts.xml.v8.5.2.backup&quot;.
 En renommant &quot;shortcuts.xml.v8.5.2.backup&quot; -&gt; &quot;shortcuts.xml&quot;, vos commandes devraient être proprement restaurées."/><!-- HowToReproduce: Close Notepad++, remove shortcuts.xml.v8.5.2.backup & v852ShortcutsCompatibilityWarning.xml if present, relaunch Notepad++, delete or modify a shortcuts via Shortcut Mapper, close Notepad++, then the message will show up -->
+			<FileInaccessibleUserSession title="Fichier inaccessible" message="Certains fichiers de votre session sauvegardée sont inaccessibles. Ils peuvent être ouverts comme des fichiers vides et en lecture seule en tant que espaces réservés.
```
Comment (Mateos81) [inline PowerEditor/installer/nativeLang/french.xml:1485]: Que e -> qu'e (twice here, in 2 diffs)

**Comment by Yaron10** [comment]
@donho,

> Fixed.

👍 

But this creates another issue.
**STR:**
Open file A.
Save session as SesA.ses.
Open file B.
Move B to the other view.
Save session as SesAB.ses.
Move B to the other view (-> A and B in one view).
Close NPP.
Delete A and B.
Select the two user-sessions in File-Explorer and open them in one go.

**Result:**
No B in MAIN_VIEW.

***

> How about replacing bool loadSession...

I understand you don't like it.

***

> "Open Session in New Instance (and save on exit)" complicates the implementation/messages.

I'm referring to the potential confusion you mentioned in https://github.com/notepad-plus-plus/notepad-plus-plus/pull/14202#issuecomment-1771799125.

The messages should indeed make it clear.

**User-sessions.**
```
Some files from your manually-saved session are inaccessible. They can be opened as empty and read-only place-holders.
Would you like to create those place-holders?

NOTE: Choosing not to create the place-holders or closing them later, your manually-saved session will NOT be modified on exit.
```

**session.xml.**
```
Some files from your past session are inaccessible. They can be opened as empty and read-only place-holders.
Would you like to create those place-holders?

NOTE: Choosing not to create the place-holders or closing them later, your session WILL BE MODIFIED ON EXIT!
You can backup your \"session.xml\" now.
```

"Open Session in New Instance (and save on exit)" **will indeed cause confusion**.
Regardless, I don't understand why you should make an exception in this case and modify the user-session.

**Comment by Yaron10** [comment]
> But this creates another issue.
STR:

You get the same result without deleting the files. - Probably not related to this PR.
It should be further investigated with current master **and older NPP versions**.

**Comment by donho** [comment]
> But this creates another issue. 
**STR:** Open file A. Save session as SesA.ses. Open file B. Move B to the other view. Save session as SesAB.ses. Move B to the other view (-> A and B in one view). Close NPP. Delete A and B. Select the two user-sessions in File-Explorer and open them in one go.
> 
> **Result:** No B in MAIN_VIEW.

Cannot reproduce it:

image

>> How about replacing bool loadSession...
> I understand you don't like it.

Implementing...

**Comment by donho** [comment]
@Yaron10 
> The messages should indeed make it clear.

Done. However, according google, "place-holder" should be "placeholder", so I maintain "placeholder".

> "Open Session in New Instance (and save on exit)" will indeed cause confusion.
Regardless, I don't understand why you should make an exception in this case and modify the user-session.

"Open Session in New Instance (and save on exit)" was a frequent-request feature which was implemented quite long time ago, so I believe a lot of users depend on this feature.

You can check the latest commit in which the messages have been refined (with the user session file name).

**Comment by Yaron10** [comment]
@donho,

It seems to work like  charm. 👍 
I hope we haven't missed some corner cases.

Thanks again.

> Done. However, according google, "place-holder" should be "placeholder", so I maintain "placeholder".

👍 
We need a native English speaker.

```
They can be opened as empty and read-only placeholder files.
```
Instead of
```
They can be opened as empty and read-only files as placeholders.
```
?

And `\n` before `You can backup`?

***

> "Open Session in New Instance (and save on exit)" was a frequent-request feature which was implemented quite long time ago,

Do you mean the **(and save on exit)** part was a frequent-request feature?

***

> You get the same result without deleting the files. - Probably not related to this PR.
It should be further investigated with current master **and older NPP versions**.

Could you please try the _exact steps_?

**STR:**
Open file A.
Save session as SesA.ses.
Open file B.
Move B to the other view.
Save session as SesAB.ses.
Move B to the other view (**-> A and B in one view on exit**).
Close NPP.
Select the two user-sessions in File-Explorer and open them in one go.

**Result:**
No B in MAIN_VIEW.

**Comment by Yaron10** [comment]
@alankilborn,

Could you please test this PR?
Thank you.

**Comment by pryrt** [comment]
Regarding "place-holder" vs "placeholder", I definitely agree that "placeholder" is the right choice.

@alankilborn suggested I chime in, since I was the one who submitted #12079.  Based on my experiments, this meets my expectations for reasonable behavior regarding the session when a file/directory/mounted-drive goes temporarily missing, and I would be happy if this is the version that goes "live".

Thank you for all the effort that the whole team put into going through various edge cases, and all the back-and-forth that occurred.  The end result shows the benefit of getting a lot of different perspectives on this issue.

**Comment by Yaron10** [comment]
@pryrt,

Thank you for submitting the issue and testing this PR.

@alankilborn,

That was a nice dodge. :)

**Comment by donho** [comment]
@Yaron10 
I do believe "They can be opened as empty and read-only files **as** placeholders" since it's chatgpt has suggested me :)
However, I think to use "documents" instead of "files" is even better: "They can be opened as empty and read-only **documents** as placeholders".

> Do you mean the (and save on exit) part was a frequent-request feature?

Yes.

> Could you please try the exact steps?

Cannot always reproduce it.

image

Note that I copied new compilerf bin to replace `Path : C:\Program Files\Notepad++\notepad++.exe` and I used `Edit with Notepad++` to open both session A.sess & AB.sess.

**Comment by Yaron10** [comment]
@donho,

> I do believe "They can be opened as...

👍 

***

> and I used Edit with Notepad++ to open both session A.sess & AB.sess.

I'm using NPP portable, and I don't have that entry in the context menu.

Both files are selected, but does it make a difference if you right-click on A.sess or AB.sess?
_And if you use Open?_

I'm using [QTTabBar](http://qttabbar.wikidot.com/) and [OldNewExplorer](https://www.softpedia.com/get/Tweak/System-Tweak/OldNewExplorer.shtml) on Windows 10.

On using Open in the bar
תמונה

I always get a wrong result: No B in MAIN_VIEW.

On using Open in the context menu, it depends on the file I right-click.

_https://github.com/notepad-plus-plus/notepad-plus-plus/issues/14161 might be related._

**Comment by donho** [comment]
@Yaron10 
Thank you for the more detail info

> Both files are selected, but does it make a difference if you right-click on A.sess or AB.sess?

I suppose by "right-click" you means **Edit with Notepad++** of context menu under Explorer which comes Windows 10/11. 
When you right-click on A.sess or AB.sess and run **Edit with Notepad++** from the context menu, NppShell.dll runs command `notepad++.exe "A.sess" "AB.sess"` - both session files are opened in one call, and session files are opened one after another.

In this scenario, it works fin, as expected.

> And if you use Open?

Always under Explorer which comes Windows 10/11, when you select 2 files and use Open command from the context menu, or type ENTER key to open 2 files:  the system run `notepad++.exe "A.sess"` then `notepad++.exe "AB.sess"` (or in inverse order, it depends in which order you select the 2 files).

In this scenario, I can reproduce the problem, not exactly the same result though:

image

I will see what I can do about it.

**Comment by donho** [comment]
> I will see what I can do about it.

The test results (with **open**) are different on mono-instance - multi-instance works perfectly - and it's surely due to the race condition: System launches several Notepad++ commands simultaneously and all commands lead to one instance for the execution in a random order.
Unfortunately I've no solution so far. And at the same time, this issue is related to this feature but it's not a regression due to this PR, so I will merge this PR.

**Comment by Yaron10** [comment]
@donho,


> In this scenario, I can reproduce the problem, not exactly the same result though (but even worse):

Yes, I encountered a similar case after posting my STR.

> I will see what I can do about it.

Thank you for looking into it and good luck.
Could you pleas open an issue? - Your specs are (obviously) much more common.

> And at the same time, this issue is related to this feature but it's not a regression due to this PR, so I will merge this PR.

👍

**Comment by donho** [comment]
@Yaron10 
> Could you please open an issue? - Your specs are (obviously) much more common.

It's indeed the same issue of #14161: 
I tested with 1 modified untitled text file in the session. With no Notepad++ open, I select 4 text files and launch **Open** via Explorer.
After several tests I cannot reproduce the lost of the modified untitled document in the current session as @OsakaWebbie has experienced. However, I have reproduced once (among 6-8 times) that 1 of these 4 files was missing - it's what it has happened to you with session files.
As I have explain [here ](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/14252#issuecomment-1774152943) & [there](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/14161#issuecomment-1774128291), system launches several commands to call Notepad++, that will raise the race condition (probably it's [better explained by portableapps](https://portableapps.com/comment/254940#comment-254940)).

As a result, I won't open a similar issue as #14161, and this issue won't be fixed and probably can't be fixed.
The best practice is using the "Edit with Notepad++" of context menu or a one command with all files as arguments in command line.

**Comment by Yaron10** [comment]
@donho,

First, I'd like to emphasize again that I was initially wrong: the issue is not related to this PR, and can be reproduced in v8.5.7 as well.

> In this scenario, I can reproduce the problem, not exactly the same result though (but even worse)

I understand your explanation and recommendation.
But ending up with two identical-path tabs in the same view is a major problem, isn't it?
(And again, it can happen with **accessible files** as well).

The question is if NPP should (and could) make sure this is not the end result.
If no other solution is viable - how about checking if an identical-path already exists in the view, and prevent opening another one?

**Comment by Yaron10** [comment]
Alternatively, and as it's a rare case, it can be labeled as a "known issue".
The best applications have them. :)

**Comment by donho** [comment]
> If no other solution is viable - how about checking if an identical-path already exists in the view, and prevent opening another one?

It's already done in the code.
The problem of race condition, that is the same resource is accessed at the same time, and identical-path checking for 2 threads or 2 processes have the same result. That's the reason that the identical-path checking not working in my test.

> Alternatively, and as it's a rare case, it can be labeled as a "known issue".

Why not? I can always labeled #14161 as "known issue". But what's the advantage?

**Comment by Yaron10** [comment]
Thank you for the explanation.

> I can always labeled https://github.com/notepad-plus-plus/notepad-plus-plus/issues/14161 as "known issue".

That issue is only reproducible with PortableApps.
The cases we've encountered her are valid in common regular setups. 

> But what's the advantage?

So that we at least have an open issue and not just some comments in a closed PR.
As you understand. :)

**Pull request closed without merging**